source/slang/slang-options.cpp

```diff
@@ -1616,6 +1619,12 @@ SlangResult OptionsParser::_parseProfile(const CommandLineArg& arg)
         }
     }
 
+    // Hack. Library profile targets imply "whole program" compilation
+    if (profileName == "lib_6_6")
```

Comment:
I had an issue where the library profile specified explicitly by the user was overridden implicitly when an entrypoint was discovered. This is my workaround.  Please advise if there's a preferred way to prevent slang from overriding lib* profiles in the presence of entry points.

Reply:
My guess is that since `lib_6_6` is not an actual capability what you were experiencing was `lib_6_6` becoming `sm_6_6`? Apologies if this answer is not exact enough, I don't exactly understand what you mean by "overridden implicitly" since any overridden capability should(?) warn or assert.

As a suggestion to maybe avoid the "are you lib_x_y" code is that we could add to our `.capdef` file a new capability: `def hlsl_lib : hlsl`. 

This way if a compile needs to compile an "hlsl lib", all that needs to be checked is:
1. does your compile capabilities for your target contain the `hlsl_lib` atom.
2. what is your current `sm_x_y` version.

Reply:
Without the whole program option, slang overrides the explicitly chosen profile and compiles it as cs_6_6 instead since it discovers the computeMain entrypoint.

Using your suggestion, what would the command line change to in the test?

`TEST(windows):COMPILE: tests/library/precompiled-dxil.slang -o tests/library/precompiled-dxil.slang-module -embed-dxil -profile lib_6_6 -incomplete-library`

Reply:
My gut feeling is that it would still be a hack even if I added all the lib versions. Of course, if this is considered a fine approach, I need to add the rest of the versions.  I'm curious to understand Ariel's suggestion.

Reply:
I am wondering if this is actually true for all the time.
> Library profile targets imply "whole program" compilation

Currently, you can use "lib_6_6" as a profile with an option to compile the whole program explicitly.
If the assumption is true that "lib_x_y" should always compile with "whole program" option, we should make this line less hacky.

Reply:
> Using your suggestion, what would the command line change to in the test?

Nothing would have to change

The idea is:
if we find a `-embed-dxil` parameter we will automatically add `hlsl_lib` to the capabilities of our compile. If the capabilities are then invalid, we know that the target is not compatible with dxil libraries. This will also remove the need to check for `lib_6_6` explicitly, only the `hlsl_lib` capability atom will need to be checked for.

`hlsl_lib` will be defined as `def hlsl_lib : _sm_6_6`. This will ensure that if any capability does not conform to a minimum of `hlsl + sm_6_6` we will error.

Reply:
Note: due to how capabilities work we can treat `hlsl_lib+sm_6_7` as valid whilst `hlsl_lib+sm_6_5" is not with a little bit of checking. This means that we use the capability system to manage all of the "are we at a minimum sm_6_6".

Reply:
I would suggest us using `dxil_lib` as the profile name.

Reply:
> if we find a -embed-dxil parameter we will automatically add/check if our capabilities are a super set of dxil_lib. If our capabilities are not a super set then we know that the compile is not compatible with dxil libraries. This will also remove the need to check for lib_x_y explicitly since only the dxil_lib capability will needed to be checked for to determine conformance.
> 
> dxil_lib will be defined as alias dxil_lib : _sm_6_6. This will ensure that if any capability 'XYZ' does not conform to a minimum of hlsl + sm_6_6 ("conform to a minimum" meaning 'XYZ' is a super-set) we will diagnose and error accordingly for the user.

The goal of checking lib_6_6 here wasn't to ensure we had at least lib_6_6, but rather just to set the GenerateWholeProgram option. AFAICT the actual minimum needed is lib_6_3, so I created a capability alias for _sm_6_3 and added it to the raw target request I created.

I'm missing something though because the capability I added does not cause any new validation error when my test specifies a too-low profile number.  Like if the test specifies "lib_6_1" I just get an error from dxc.

Reply:
> I'm missing something though because the capability I added does not cause any new validation error when my test specifies a too-low profile number. Like if the test specifies "lib_6_1" I just get an error from dxc.

I think I gave partially misleading advice, I will leave a new code review comment describing what to change.

To give a general answer though of why there are no errors, the end goal is to check if `cookedCapabilities.implies(CapabilitySet(dxil_lib))` is false for when `-embed-dxil` is true.